src/types/num.rs

```diff
@@ -242,10 +232,175 @@ int_convert_u64_or_i64!(
     ffi::PyLong_AsUnsignedLongLong
 );
 
+#[cfg(all(not(Py_LIMITED_API), target_endian = "little"))]
+int_convert_128!(i128, 16, 1);
 #[cfg(not(Py_LIMITED_API))]
-int_convert_bignum!(i128, 16, IS_LITTLE_ENDIAN, 1);
-#[cfg(not(Py_LIMITED_API))]
-int_convert_bignum!(u128, 16, IS_LITTLE_ENDIAN, 0);
+int_convert_128!(u128, 16, 0);
+
+#[cfg(all(feature = "num-bigint", not(Py_LIMITED_API)))]
+mod bitint_conversion {
+    use super::*;
+    use num_bigint::{BigInt, BigUint};
+
+    unsafe fn extract_small(ob: &PyAny, n: usize, is_signed: c_int) -> PyResult<[c_uchar; 128]> {
+        let buffer = [0; 128];
+        let ok = ffi::_PyLong_AsByteArray(
+            ob.as_ptr() as *mut ffi::PyLongObject,
+            buffer.as_ptr() as *const c_uchar,
+            n,
+            1,
+            is_signed,
+        );
+        if ok == -1 {
+            Err(PyErr::fetch(ob.py()))
+        } else {
+            Ok(buffer)
+        }
+    }
+
+    unsafe fn extract_large(ob: &PyAny, n: usize, is_signed: c_int) -> PyResult<Vec<c_uchar>> {
+        let buffer = vec![0; n];
+        let ok = ffi::_PyLong_AsByteArray(
+            ob.as_ptr() as *mut ffi::PyLongObject,
+            buffer.as_ptr() as *const c_uchar,
+            n,
+            1,
+            is_signed,
+        );
+        if ok == -1 {
+            Err(PyErr::fetch(ob.py()))
+        } else {
+            Ok(buffer)
+        }
+    }
+
+    macro_rules! bigint_conversion {
+        ($rust_ty: ty, $is_signed: expr, $to_bytes: path, $from_bytes: path) => {
+            impl ToPyObject for $rust_ty {
+                fn to_object(&self, py: Python) -> PyObject {
+                    unsafe {
+                        let bytes = $to_bytes(self);
+                        let obj = ffi::_PyLong_FromByteArray(
+                            bytes.as_ptr() as *const c_uchar,
+                            bytes.len(),
+                            1,
+                            $is_signed,
+                        );
+                        PyObject::from_owned_ptr_or_panic(py, obj)
+                    }
+                }
+            }
+            impl<'source> FromPyObject<'source> for $rust_ty {
+                fn extract(ob: &'source PyAny) -> PyResult<$rust_ty> {
+                    unsafe {
+                        let num = ffi::PyNumber_Index(ob.as_ptr());
+                        if num.is_null() {
+                            return Err(PyErr::fetch(ob.py()));
+                        }
+                        let n_bytes = (ffi::_PyLong_NumBits(ob.as_ptr()) as usize - 1) / 8 + 1;
```

Comment:
looks like this should use `num`, not `ob`...

Can `_PyLong_NumBits` return zero? The subtract would wrap in that case.

Reply:
> looks like this should use num, not ob...

Wow, I didn't know that.

> Can _PyLong_NumBits return zero? The subtract would wrap in that case.

Good catch, thanks.